src/Amazonka/S3/StreamingUpload.hs

```diff
@@ -283,34 +284,43 @@ data S = S !Builder {-# UNPACK #-} !Int
 newS :: S
 newS = S mempty 0
 
-newSFrom :: ByteString -> S
-newSFrom bs = S (byteStringCopy bs) (B.length bs)
-
 appendS :: S -> ByteString -> S
 appendS (S builder len) bs = S (builder <> byteStringCopy bs) (len + B.length bs)
 
 finaliseS :: S -> IO ByteString
 finaliseS (S builder builderLen) = do
   fptr <- mallocForeignPtrBytes builderLen
   let ptr = unsafeForeignPtrToPtr fptr
-      bufWriter = runBuilder builder
-  bufWriter ptr builderLen >>= \case
+  runBuilder builder ptr builderLen >>= \case
     (written, Done)
       | written == builderLen -> pure $! PS fptr 0 builderLen
       | otherwise ->
           error $ "finaliseS: bytes written didn't match, expected: " <> show builderLen <> " got: " <> show written
     (_written, _) -> error "Something went very wrong"
 
--- Right means the buffer needs more data to fill it
--- Left means the buffer is full
-processChunk :: ChunkSize -> ByteString -> S -> IO (Either S S)
+-- @Right@ means the buffer needs more data to fill it
+-- @Left@ means the buffer is full so it should be yielded, and returns the
```

Comment:
Suggestion: make a custom type whose constructors contain the meaning here.
Bonus: you can add strictness annotations to the constructors where they make sense.